Drop CachePayload::data() in favor of span()

cdumez opened the pull request:
#### 098c58dd13bf40fc81971361162e21d05cb1f74a
<pre>
Drop CachePayload::data() in favor of span()
<a href="https://bugs.webkit.org/show_bug.cgi?id=276051">https://bugs.webkit.org/show_bug.cgi?id=276051</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/CachePayload.h:
* Source/JavaScriptCore/runtime/CachedBytecode.h:
(JSC::CachedBytecode::span const):
(JSC::CachedBytecode::data const): Deleted.
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::Decoder::offsetOf):
(JSC::Decoder::ptrForOffsetFromBase):
(JSC::decodeCodeBlockImpl):
(JSC::isCachedBytecodeStillValid):

Canonical link: <a href="https://commits.webkit.org/280565@main">https://commits.webkit.org/280565@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/98a0c04b1a14fb9d548ebcd878a74acbb83468df

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/56986 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/36314 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/9461 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/60606 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/7429 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/43937 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/7619 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/46151 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/5220 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/59016 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/34101 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/49185 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27011 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/56512 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/30881 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/6516 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/6434 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/50078 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/52839 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/6787 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/62287 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/56227 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/899 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/6891 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/53410 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/902 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/49240 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/53447 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/762 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/77988 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8489 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/32143 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12928 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/33228 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/34313 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/32974 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/CachedTypes.cpp:
```diff
@@ -304,8 +304,9 @@ size_t Decoder::size() const
-    ASSERT(addr >= m_cachedBytecode->data() && addr < m_cachedBytecode->data() + m_cachedBytecode->size());
-    return addr - m_cachedBytecode->data();
+    auto cachedBytecodeSpan = m_cachedBytecode->span();
+    ASSERT(addr >= cachedBytecodeSpan.data() && addr < (cachedBytecodeSpan.data() + cachedBytecodeSpan.size()));
```
You'd think there would be an idiomatic way to write this. Or should we add a helper function if the standard library doesn't have it?

cdumez replied:
Maybe I should add something like this in StdLibExtras.h:
```
bool isInSpanRange(std::span<T> span, T* address);
```

darinadler replied:
If there isn't something like that in the standard library, I think it would be good. I'd want it to work even when the const-ness of the pointer and the span differed. Maybe the ranges part of the standard library has this.

darinadler replied:
I think it's ambiguous whether a point is in a range of it's at the end. That makes this function name annoying.

cdumez replied:
Ok, if it's confusing then I'm going back to not having this function. It is only used in one place at the moment anyway.

cdumez committed:
Drop CachePayload::data() in favor of span()
https://bugs.webkit.org/show_bug.cgi?id=276051

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/CachePayload.h:
* Source/JavaScriptCore/runtime/CachedBytecode.h:
(JSC::CachedBytecode::span const):
(JSC::CachedBytecode::data const): Deleted.
* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::Decoder::offsetOf):
(JSC::Decoder::ptrForOffsetFromBase):
(JSC::decodeCodeBlockImpl):
(JSC::isCachedBytecodeStillValid):

Canonical link: https://commits.webkit.org/280565@main
